src/main/java/com/google/devtools/build/lib/skyframe/ProcessPackageDirectory.java

```diff
@@ -187,6 +182,47 @@ public ProcessPackageDirectoryResult getPackageExistenceAndSubdirDeps(
         /*additionalValuesToAggregate=*/ ImmutableMap.of());
   }
 
+  // Note that it's not enough to just check for the convenience symlinks themselves,
+  // because if the value of --symlink_prefix changes, the old symlinks are left in place. It
+  // is also not sufficient to check whether the symlink points to a directory in the current
+  // exec root, since this can change between bazel invocations. Therefore we check if the
+  // suffix of the symlink source suggests it is a convenience symlink, then see if the symlink
+  // target is in a directory that looks like an execroot. This algorithm also covers more
+  // creative use cases where people create convenience symlinks somewhere in the directory
+  // tree manually.
+  private boolean isConvenienceSymlink(FileValue fileValue, RootedPath rootedPath, SkyFunction.Environment env) throws InterruptedException {
+    if (!fileValue.isSymlink()) {
+      return false;
+    }
+
+    PathFragment rootRelativePath = rootedPath.getRootRelativePath();
+    Root root = rootedPath.getRoot();
+    PathFragment linkTarget = fileValue.getUnresolvedLinkTarget();
+
+    return linkTarget.startsWith(directories.getExecRootBase().asFragment())
+    || (rootRelativePath.getBaseName().endsWith("-bin") && isInExecRoot(linkTarget, root, 4, env))
+    || (rootRelativePath.getBaseName().endsWith("-out") && isInExecRoot(linkTarget, root, 2, env))
+    || (rootRelativePath.getBaseName().endsWith("-testlogs") && isInExecRoot(linkTarget, root, 4, env))
+    || (rootRelativePath.getBaseName().endsWith("-" + directories.getWorkingDirectory().getBaseName()) && isInExecRoot(linkTarget, root, 1, env));
+  }
+
+  private boolean isInExecRoot(PathFragment path, Root root, int depth, SkyFunction.Environment env) throws InterruptedException {
+    PathFragment candidateExecRoot = path.subFragment(0, path.segmentCount() - depth);
+
+    if(!candidateExecRoot.getBaseName().equals("execroot")) {
+      return false;
+    }
+
+    Root absoluteRoot = Root.absoluteRoot(root.getFileSystem());
+    RootedPath doNotBuildPath = RootedPath.toRootedPath(absoluteRoot, candidateExecRoot.getChild("DO_NOT_BUILD_HERE"));
```

Comment:
I would like to use `BlazeWorkspace.DO_NOT_BUILD_FILE_NAME` here, but doing so introduces a circular dependency:

```
.-> //src/main/java/com/google/devtools/build/lib:runtime (ad2a6540989c55656a83c238ad8764286067c713283fa652a29a1eb44476b1c2)
|   //src/main/java/com/google/devtools/build/lib/skyframe:skyframe_cluster (ad2a6540989c55656a83c238ad8764286067c713283fa652a29a1eb44476b1c2)
|   //src/main/java/com/google/devtools/build/lib/skyframe:recursive_pkg_function (ad2a6540989c55656a83c238ad8764286067c713283fa652a29a1eb44476b1c2)
|   //src/main/java/com/google/devtools/build/lib/skyframe:recursive_directory_traversal_function (ad2a6540989c55656a83c238ad8764286067c713283fa652a29a1eb44476b1c2)
|   //src/main/java/com/google/devtools/build/lib/skyframe:process_package_directory (ad2a6540989c55656a83c238ad8764286067c713283fa652a29a1eb44476b1c2)
`-- //src/main/java/com/google/devtools/build/lib:runtime (ad2a6540989c55656a83c238ad8764286067c713283fa652a29a1eb44476b1c2)
```